docs/custom-filters.adoc

```diff
@@ -45,23 +157,173 @@ Instances are specific to the channel between a single client and a single broke
 
 It exists while the client remains connected.
 
+=== Filter Patterns
+
+Kroxylicious Protocol Filters support several patterns:
+
+1. <<Synchronously intercepting/mutating Requests and Responses>>
+2. <<Responding from a Request Filter without Proxying>>
+3. <<Sending new messages to Kafka that are not visible to the Client>>
+4. <<Filtering specific API Versions>>
+
+==== Synchronously intercepting/mutating Requests and Responses
+
+This is a common pattern, we want to inspect or modify a message:
+
+[source,java]
+----
+public class SampleFetchResponseFilter implements FetchResponseFilter {
+    @Override
+    public void onFetchResponse(short apiVersion,
+                                ResponseHeaderData header,
+                                FetchResponseData response,
+                                KrpcFilterContext context) {
+        mutateResponse(response, context);
+        context.forwardResponse(header, response);
+    }
+}
+----
+
+==== Responding from a Request Filter without Proxying
+
+In some cases we may wish to not forward a message from the client to Kafka. Instead, we want to generate a response
+message in Kroxylicious and send it back to the client. For example:
+
+[source,java]
+----
+public class ApiVersionsErrorFilter implements ApiVersionsRequestFilter {
+
+    @Override
+    public void onApiVersionsRequest(short apiVersion,
+                                     RequestHeaderData header,
+                                     ApiVersionsRequestData request,
+                                     KrpcFilterContext context) {
+        ApiVersionsResponseData response = new ApiVersionsResponseData();
+        response.setErrorCode(Errors.UNKNOWN_SERVER_ERROR.code());
+        context.forwardResponse(response);
+    }
+}
+----
+This will respond to all Api Versions requests with an error response without forwarding any of those requests to Kafka.
+
+Again, `forwardResponse` must be called synchronously, within the `onXyzRequest` method.
+
+==== Sending new messages to Kafka that are not visible to the Client
+
+Filters can also send arbitrary new messages to Kafka that will be invisible to the client. For example we could sent a
```

Comment:
Again, you need to get across that the request will traverse the filter chain that exists to the upstream of this filter.

Reply:
Have added more explanation.

I had a go at diagramming this one too but it got too messy (there is the main request plus the out-of-band request in play, and then there's the fact that upstream Request filters are invoked for the out-of-band request but response filters do not get invoked for the corresponding Response).

Reply:
At first glance that sounds like a big, why is it only traversing the filter chain one way?

Reply:
I should confirm that behaviour, it is indeed a bit interesting that it's one way only.

Reply:
I confirmed it [experimentally](https://github.com/kroxylicious/kroxylicious/compare/main...robobario:kproxy:intercept-out-of-band). Will make an issue to find the right solution separately.